pkg/daemon/controller/controller.go

```diff
@@ -203,6 +203,130 @@ func (sc *ServiceController) Reconcile(_ context.Context, req ctrl.Request) (res
 	return ctrl.Result{}, nil
 }
 
+// Cleanup removes an instance from the cluster. If the upgradeInPlace flag is true, the node object is not deleted.
+// TODO: Remove/ensure files does not exist and also remove HNS networks, like WMCO does through windows.Deconfigure()
+func (sc *ServiceController) Cleanup(upgradeInPlace bool) error {
+	klog.Infof("cleaning up %s, node upgrade=%t", sc.nodeName, upgradeInPlace)
+	var node core.Node
+	err := sc.client.Get(sc.ctx, client.ObjectKey{Name: sc.nodeName}, &node)
+	if err != nil {
+		return err
+	}
+
+	desiredVersion, present := node.Annotations[nodeconfig.DesiredVersionAnnotation]
+	if !present {
+		return errors.Errorf("node missing desired version annotation")
+	}
+	// Fetch the CM of the desired version
+	cm := &core.ConfigMap{}
+	err = sc.client.Get(sc.ctx,
+		client.ObjectKey{Namespace: sc.watchNamespace, Name: servicescm.NamePrefix + desiredVersion}, cm)
+	if err != nil {
+		return err
+	}
+	return sc.removeServices(cm, node, upgradeInPlace)
+}
+
+// removeServices removes all the services defined in the given ConfigMap from this instance.
+// The given Node object is deleted if the upgrade flag is false
+func (sc *ServiceController) removeServices(cm *core.ConfigMap, node core.Node, upgradeInPlace bool) error {
+	klog.Infof("removing services defined in ConfigMap %s", cm.Name)
+	cmData, err := servicescm.Parse(cm.Data)
+	if err != nil {
+		return err
+	}
+	for _, service := range cmData.Services {
+		if err := sc.ensureServiceIsRemoved(service.Name); err != nil {
+			return err
+		}
+	}
+
+	if !upgradeInPlace {
+		if err = sc.client.Delete(sc.ctx, &node); err != nil {
+			return errors.Wrapf(err, "error deleting node %s", node.GetName())
+		}
+		klog.Infof("deleted node %s", node.GetName())
+	}
+	return nil
+}
+
+// ensureServiceIsRemoved deletes the given service and all services that depend on it, even those unmanaged by WICD.
+func (sc *ServiceController) ensureServiceIsRemoved(service string) error {
+	existingSvcs, err := sc.getExistingServices()
+	if err != nil {
+		return errors.Wrap(err, "could not determine existing Windows services")
+	}
+	// Nothing to do if it already does not exist
+	if _, present := existingSvcs[service]; !present {
```

Comment:
Is this a genuinely breaking statement? No other checks are required if the `service` is not in the list of existing services. Correct?

Reply:
Yep, that's correct. If a service does not exist, there's no way to clean it up further.

Reply:
Makes sense.